Hacky attempt at AWS Provider patching to help users work around Terraform bug

brikis98 opened the pull request:
This PR adds an `aws-provider-patch` command that can be used to override attributes in nested `provider` blocks. This is an attempt at a hacky workaround for [a Terraform bug](https://github.com/hashicorp/terraform/issues/13018) where `import` does not work if you are using any modules that have `provider` blocks nested in within them. With this PR, users could run:

```bash
terragrunt aws-provider-patch --terragrunt-override-attr region=eu-west-1
```

And Terragrunt will:

1. Run `terraform init` to download the code for all your modules into `.terraform/modules`.
1. Scan all the Terraform code in `.terraform/modules`, find AWS `provider` blocks, and hard-code the `region` param
   to `eu-west-1` for each one. 

Once you do this, you'll hopefully be able to run `import` on that module. After that, you can delete the modified `.terraform/modules` and go back to normal.

Note that I wanted to document this command somewhere, so I updated the CLI docs to include all Terragrunt commands, and not just options. Therefore, as a side effect, this PR also fixes #1306.

brikis98 commented:
I got the patching code working, but there's a bit of an issue: Terraform downloads modules into `.terraform/modules`, but it downloads the _entire_ repo into that folder, and not just a single module... So the patching code either:

1. Updates every file in the entire repo, which is wasteful, and potentially results in parsing files that perhaps shouldn't be parsed.
1. We have to parse `.terraform/modules/modules.json` to know exactly in which folders to look... But I'm not sure if that's a stable API we can rely on.

yorinasub17 commented on cli/aws_provider_patch.go:
```diff
@@ -0,0 +1,131 @@
+// the .terraform/modules folder; it does NOT return Terraform files for the top-level (AKA "root") module.
+//
+// NOTE: this method only supports *.tf files right now. Terraform code defined in *.json files is not currently
+// supported.d
```
```suggestion
// supported.
```

brikis98 replied:
Fixed: https://github.com/gruntwork-io/terragrunt/pull/1326/commits/0db564ace59f9ea140363172ee614d694e5ed824

yorinasub17 commented on cli/aws_provider_patch.go:
```diff
@@ -0,0 +1,131 @@
+		return nil, nil
+	}
+
+	// Ideally, we'd use a builin Go library like filepath.Glob here, but per https://github.com/golang/go/issues/11862,
```
```suggestion
	// Ideally, we'd use a builtin Go library like filepath.Glob here, but per https://github.com/golang/go/issues/11862,
```

brikis98 replied:
Fixed: https://github.com/gruntwork-io/terragrunt/pull/1326/commits/0db564ace59f9ea140363172ee614d694e5ed824

yorinasub17 commented on cli/aws_provider_patch.go:
```diff
@@ -0,0 +1,131 @@
+
+	for _, block := range hclFile.Body().Blocks() {
+		tokens := block.BuildTokens(nil)
+		if len(tokens) > 4 {
```
If you switch to the newer HCL2 library (`"github.com/hashicorp/hcl/v2"`), you should have access to `block.Type()` and `block.Labels`, which should simplify this to:

```
if block.Type() == "provider" && len(block.Labels()) == 1 && block.Labels()[0] == "aws" {
```

brikis98 replied:
Whhhhhaaaat? Will try now, thanks! 😁

brikis98 replied:
Well goddamn. It worked. Nice one Yori! https://github.com/gruntwork-io/terragrunt/pull/1326/commits/bdd6eed5b941e31a365774ca0bf40170f4c92f90

yorinasub17 commented on cli/aws_provider_patch.go:
```diff
@@ -0,0 +1,179 @@
+	// into that folder, but the entire repo it's in, which can contain lots of other unrelated code we probably don't
+	// want to touch. To find the paths to the actual modules, we read the modules.json file in that folder, which is
+	// a manifest file Terraform uses to track where the modules are within each repo. Note that this is an internal
+	// API, so the way we parse/read this modules.json file may break in future Terraform versions.
```
Add link to https://github.com/hashicorp/terraform/blob/master/internal/modsdir/manifest.go as a reference (Note that we can't import that directly because it is labeled as internal).

brikis98 replied:
Good idea: https://github.com/gruntwork-io/terragrunt/pull/1326/commits/8fcc5060861ec81ccf754f81fb9c79248629fa7f

brikis98 commented on cli/aws_provider_patch.go:
```diff
@@ -0,0 +1,131 @@
+	"github.com/gruntwork-io/terragrunt/util"
+	"github.com/hashicorp/hcl2/hcl"
+	"github.com/hashicorp/hcl2/hcl/hclsyntax"
+	"github.com/hashicorp/hcl2/hclwrite"
```
Fixed: https://github.com/gruntwork-io/terragrunt/pull/1326/commits/bdd6eed5b941e31a365774ca0bf40170f4c92f90

brikis98 commented:
Update: I decided to bite the bullet and parse the `.terraform/modules/modules.json` format after all. Not sure of a better approach.

Note that I wanted to document this command somewhere, so I updated the CLI docs to include all Terragrunt commands, and not just options. Therefore, as a side effect, this PR also fixes #1306. I've updated the description of this PR as well, and I believe it's now ready for review.

yorinasub17 commented on test/integration_test.go:
```diff
@@ -1214,6 +1215,18 @@ func TestAutoRetryApplyAllDependentModuleRetries(t *testing.T) {
+	modulePath := util.JoinPath(rootPath, TEST_FIXTURE_AWS_PROVIDER_PATCH)
+	err := runTerragruntCommand(t, fmt.Sprintf("terragrunt aws-provider-patch --terragrunt-override-attr region=eu-west-1 --terragrunt-working-dir %s", modulePath), os.Stdout, stderr)
+
+	assert.Nil(t, err)
```
NIT: Use `assert.NoError`. The difference is that when there is an error, this will render the error object so you can see the error message (as opposed to some weird pointer).

brikis98 replied:
Good call! https://github.com/gruntwork-io/terragrunt/pull/1326/commits/0fc46e35e5eb595dfedf5ddbd8c1b944653b22f7

yorinasub17 reviewed (dismissed):
LGTM!

brikis98 commented:
Can I haz one more review plz? 😄

brikis98 commented:
Thanks for the review! Merging now.
